attestation-evidence-broker/src/aeb_modules/sev_aeb/get_evidence.rs

```diff
@@ -0,0 +1,144 @@
+                    ));
+                    rsp = reqwest::blocking::get(url);
+                }
+            }
+            // HTTP request has failed.
+            Err(_) => break,
+        }
+    }
+    let mut rsp = rsp.context(format!(
+        "Failed to complete request: {:?}\nError codes received from server:\n{}",
+        usage,
+        http_request_replies.join("\n")
+    ))?;
+
+    let mut buf = Vec::new();
+    rsp.copy_to(&mut buf)
+        .context(format!("unable to complete {:?} download", usage))?;
+
+    Certificate::decode(&mut &buf[..], ())
+        .context(format!("unable to parse downloaded {:?}", usage))
+}
+
+fn firmware() -> Result<Firmware> {
+    Firmware::open().context("unable to open /dev/sev")
+}
+
+fn chain() -> Result<Chain> {
+    const CEK_SVC: &str = "https://kdsintf.amd.com/cek/id";
+
+    let mut chain = firmware()?
+        .pdh_cert_export()
+        .map_err(|e| Error::new(ErrorKind::Other, format!("{:?}", e)))
+        .context("unable to export SEV certificates")?;
+
+    let id = firmware()?
+        .get_identifier()
+        .map_err(|e| Error::new(ErrorKind::Other, format!("{:?}", e)))
+        .context("error fetching identifier")?;
+    let url = format!("{}/{}", CEK_SVC, id);
+
+    chain.cek = download(&url, Usage::CEK)?;
+
+    Ok(chain)
+}
+
+fn report(guest_handle: u32) -> Result<AttestationReport> {
+    let mut monce = [0_u8; 16];
+    rand::thread_rng().fill(&mut monce);
+
+    let report = firmware()?
+        .get_attestation_report(guest_handle, monce)
+        .map_err(|e| Error::new(ErrorKind::Other, format!("{:?}", e)))
+        .context("unable to get SEV guest attestation report")?;
+
+    Ok(report)
+}
+
+pub fn collect_evidence(guest_handle: u32) -> Result<SevEvidence> {
+    let cert_chain = chain()?;
+    let attestation_report = report(guest_handle)?;
+
+    Ok(SevEvidence {
+        report: attestation_report,
+        cek: cert_chain.cek,
+        pek: cert_chain.pek,
+        oca: cert_chain.oca,
+    })
+}
+
+fn _ca_chain_builtin(chain: &Chain) -> Result<ca::Chain> {
+    use std::convert::TryFrom;
+
+    Generation::try_from(chain)
+        .map_err(|_| {
+            Error::new(
+                ErrorKind::NotFound,
+                "could not find a matching builtin certificate",
+            )
+        })
+        .context("failed to deduce platform generation")
+        .map(|g| g.into())
+}
+
+#[cfg(test)]
+mod tests {
+    use self::sev::certs::Verifiable;
+    use super::*;
+
+    #[test]
+    fn test_collect_and_verify_evidence() {
+        let chain = chain().unwrap();
+        let ca_chain = _ca_chain_builtin(&chain).unwrap();
+
+        let ask = ca_chain.ask;
+        let ark = ca_chain.ark;
+
+        (&ark, &ark).verify().unwrap();
+        (&ark, &ask).verify().unwrap();
+
+        // The guest handle usually increments from 1
+        let guest_handle = 1;
+        let evidence = collect_evidence(guest_handle).unwrap();
```

Comment:
Maybe we can use `assert_eq!()` here to judge whether the obtained evidence is what we expect?